### ISSUE TITLE
[red-knot] Support narrowing on attribute and subscript expressions

### ISSUE THREAD

**Issue by dhruvmanila** [astral-sh/ruff#16894]
Currently, red-knot only supports narrowing on a name expression like `x` but not on attributes (`foo.x`) or subscripts (`foo[0]`).

For example the following is a false negative as detected in https://github.com/astral-sh/ruff/pull/16888#issuecomment-2742772151:
```py
from typing import Callable, Optional


class Config:
    formatting_function: Optional[Callable[[], None]] = None


def process(config: Config):
    reveal_type(config.formatting_function)  # revealed: () -> None | None
    if config.formatting_function:
        # red-knot: Object of type `None` is not callable [lint:call-non-callable]
        config.formatting_function()
```

**Comment by sharkdp** [comment]
@mtshiba I can't assign this to you (maybe because you haven't commented in this thread), but I understand that you are working on this. I'm assigning myself as a placeholder, so that no-one else starts with this.

I think it would be good to start with narrowing on attribute expressions, and postpone narrowing on subscript expressions to a later point in time.

**Comment by carljm** [comment]
Hi @mtshiba! Just checking in on the status here and whether there is anything that it would be useful for us to review, even if it's just a planned approach or a test suite, to make sure we are all on the same page? We are hoping we can have some form of this feature landed within the next week or two; does that seem possible?

**Comment by mtshiba** [comment]
Since I thought the fix for #17595 was relevant to this issue (though not so relevant as it turned out), I have submitted [a PR](https://github.com/astral-sh/ruff/pull/17630) to fix that one first. Please check it.

A PR to solve this issue is also currently being prepared, but the basic feature should be ready within the next week.
